drivers: eth_stm32_hal: Initial STM32F4 HAL based Ethernet driver

lowlander opened the pull request:
Signed-off-by: Erwin Rol <erwin@erwinrol.com>

erwango commented:
Hi Erwin,

I had a quick look at your patch and from HAL integration point of view,
it looks good.
Before going further in the review, is there some easy way to test it?
Did you use some sample application? (I'm not quite familiar with net
subsystem tests)

Besides, for more complete review, I think you should call for reviewers
with eth experience in zephyr, which is not my case.
@galak, any idea?

Erwan

lowlander commented:
Hey Erwan ( @erwango  ) ,

Any test of course requires hardware, and I only have my olimex_stm32_e407 board. So I am not sure how the Zephyr community deals with this, surely not every developer will have every board to test, especially when more and more boards will be supported. 

What I did was build the http_server sample, to see if things work. With that example I ran the obligatory "ICMP flood ping" test, which it passed. Also the website was shown in the browser, so the TCP connection worked. Also DHCP was used for getting the IPv4 address. So basic things work, but this is far from a "compliance" test. 

So is there any standardized way to test Ethernet drivers (or drivers in general, cause my CAN driver will have the same problem) before they will be accepted in Zephyr?

Any hints/tips are welcome. 

PS: I did notice some problem with ICMP-ping replies from Zephyr depending on if the request came from Centos 6 or Centos 7. Centos 7 seems to embed some extra checksum in the data and Wireshark recognizes that and flags the reply from Zephyr as broken. I will look into that and report it as a separate bug. 

PPS: Also I noticed something wrong with the http server, when the browser keeps the TCP connection open and sends a second request for the same page. Also will look into that and report separately.

PPPS: I have no idea how I can "call for reviewers with eth experience in zephyr".

superna9999 commented:
Hi Erwin,

Can you split the patch with all the pinmux and HAL related in separate commits ?

lowlander commented:
Hey Neil (@superna9999),

sure, just split the commit and use this same PR, or also make two PR's out of it? 

- Erwin

superna9999 commented:
Yes use the same PR

lowlander committed:
pinmux: stm32: directly return error if stm32_get_pin_config fails.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander committed:
pinmux: stm32: add ethernet controller pins

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander committed:
board: olimex_stm32_e407: update pinmux config for ethernet interface

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander committed:
driver: eth_stm32_hal: Initial STM32 HAL based Ethernet driver

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander commented:
OK noticed a typo in the last commit msg, I'll wait for the checks to finish to see if something else needs editing to and than repush.

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -0,0 +1,394 @@
+		heth->Instance->DMATPDR = 0;
+	}
+
+	net_pkt_unref(pkt);
```
The driver should only unref the pkt if it could send it properly. If there was an error, then do not unref the packet but return value < 0 so that caller can decide what to do with the packet in this case.

lowlander replied:
OK, wasn't aware of that, I will fix that.

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -0,0 +1,394 @@
+
+	/* Register Ethernet MAC Address with the upper layer */
+	net_if_set_link_addr(iface, dev_data->mac_addr,
+		sizeof(dev_data->mac_addr),
```
Indentation looks wrong here

lowlander replied:
It took several tries to make shipable happy, If you know how I should indent it without getting a shipable error I would be thankful.

jukkar replied:
That sounds really weird if shippable complains if you indent properly. Are you using tabs + spaces to indent the line, the sizeof() should start under "iface".

lowlander replied:
Retried it and now it didn't complain, maybe it was just a warning that only was shown because there also were some errors at that time. But anyway, indentation fixed.

jukkar commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -0,0 +1,394 @@
+
+static ETH_DMADescTypeDef dma_rx_desc_tab[ETH_RXBUFNB] __aligned(4);
+static ETH_DMADescTypeDef dma_tx_desc_tab[ETH_TXBUFNB] __aligned(4);
+static u8_t dma_rx_buffer[ETH_RXBUFNB][ETH_RX_BUF_SIZE] __aligned(4);
```
This uses lot of memory. Would it be possible to do this same way as eth_sam_gmac.c is doing, so mapping dma region(s) directly to net_buf data area?

lowlander replied:
The STM32 HAL (HAL_ETH_DMATxDescListInit) seems to want things like this.  And I wanted to stay as close to the HAL as possible with this driver.

jukkar commented:
> So is there any standardized way to test Ethernet drivers (or drivers in general, cause my CAN driver will have the same problem) before they will be accepted in Zephyr?

At the moment we do not have proper real hw tests available, the network tests under tests/net are suppose to be self contained and run inside qemu.

jukkar commented:
> I did notice some problem with ICMP-ping replies from Zephyr depending on if the request came from Centos 6 or Centos 7. Centos 7 seems to embed some extra checksum in the data and Wireshark recognizes that and flags the reply from Zephyr as broken. I will look into that and report it as a separate bug.

I suppose this was IPv4 ping you were doing. Have you done any IPv6 testing with this driver, there was issues with frdm_k64f ethernet driver when working with IPv6.
Did the Centos7 accept the ping reply or was it discarded in this case?

lowlander commented:
@jukkar yes IPv4, I must admit I haven't tested IPv6 at all yet. 
- Centos7 accepts all replies and doesn't complain.
- Wireshark reports the following;
"Internet Control Message Protocol
    Type: 0 (Echo (ping) reply)
    Code: 0
    Checksum: 0x8711 [incorrect, should be 0xb111]
    [Checksum Status: Bad]
    Identifier (BE): 32121 (0x7d79)
    Identifier (LE): 31101 (0x797d)
    Sequence number (BE): 3 (0x0003)
    Sequence number (LE): 768 (0x0300)
    Timestamp from icmp data: Jun 27, 2017 23:26:33.000000000 W. Europe Daylight Time
    [Timestamp from icmp data (relative): 12.308791000 seconds]
    Data (79 bytes)"
- Wireshark starts complaining when `ping -s 87` is used, which results in a 129 byte frame, which just doesn't fit in a 128 buffer anymore. This seems like a big hint.
- when I compare req and reply in wireshark only two bytes in the ICMP payload are different, the type (8=req,0=reply) and the checksum. This suggests that the Ethernet driver does recv/send the data correctly and that there is something wrong with the ICMP checksum calculation when it crosses 128byte buffer boundaries. 

I will investigate (including IPv6) more later.

jukkar commented:
> Wireshark starts complaining when ping -s 87 is used, which results in a 129 byte frame, which just doesn't fit in a 128 buffer anymore. This seems like a big hint.

As you noticed, the IPv4 version does not support larger packets. I see you sent a patch to ml, but please create a PR for that too.

jukkar commented:
> I see you sent a patch to ml, but please create a PR for that too.

I looked your proposal in ml and it looks convoluted. We have already a routing calc_chksum_pkt() that traverses the fragments. I did a quick experiment and this seems to work, can you try and verify it too.

```
diff --git a/subsys/net/ip/utils.c b/subsys/net/ip/utils.c
index f9e3e1d..f7b3115 100644
--- a/subsys/net/ip/utils.c
+++ b/subsys/net/ip/utils.c
@@ -446,7 +446,7 @@ static inline u16_t calc_chksum_pkt(u16_t sum, struct net_pkt *pkt,
 u16_t net_calc_chksum(struct net_pkt *pkt, u8_t proto)
 {
 	u16_t upper_layer_len;
-	u16_t sum;
+	u16_t sum = 0;
 
 	switch (net_pkt_family(pkt)) {
 #if defined(CONFIG_NET_IPV4)
@@ -456,11 +456,7 @@ u16_t net_calc_chksum(struct net_pkt *pkt, u8_t proto)
 			net_pkt_ipv6_ext_len(pkt) -
 			net_pkt_ip_hdr_len(pkt);
 
-		if (proto == IPPROTO_ICMP) {
-			return htons(calc_chksum(0, net_pkt_ip_data(pkt) +
-						 net_pkt_ip_hdr_len(pkt),
-						 upper_layer_len));
-		} else {
+		if (proto != IPPROTO_ICMP) {
 			sum = calc_chksum(upper_layer_len + proto,
 					  (u8_t *)&NET_IPV4_HDR(pkt)->src,
 					  2 * sizeof(struct in_addr));
```

If it works ok for you too, could you send a PR using this change?

lowlander commented:
@jukkar your fix works, I'll create a PR for it.

lowlander commented:
@jukkar are those DMA buffers really a show stopper or could it be merged and improved in a future version ?
@superna9999 do you have anymore things I should change ?

jukkar commented:
> @jukkar are those DMA buffers really a show stopper or could it be merged and improved in a future version ?

Not a show stopper, just was worried about the amount of kb used by these buffers. This can be improved in the future of course.

jukkar commented:
> I will investigate (including IPv6) more later.

I can only talk about the IP stack things (so only about one patch in this PR), but would be more confident to +1 if you could try and see if the IPv6 works with this driver. There should not be any issues as the driver should not care about what data it receives and sends. Unfortunately this was not the case for example with frdm_k64f which worked only with IPv4 by default.

lowlander commented:
@jukkar I will look into the IPv6 stuff a bit more this weekend, because yes your right the Ethernet driver should not care about IPv6 at all. So might be something like with the ICMP ping problem. I'll report back as soon as I know more.

lowlander commented:
@jukkar I found (part of) the problem, IPv6 needs multicast for neighborhood discovery. The default setting of the STM32 HAL is to perfectly filter multicast frames. If I change that to just pass all multicast frames ping6 works;

`64 bytes from fe80::280:e1ff:fe93:11e%enp5s0: icmp_seq=1185 ttl=64 time=82.7 ms`

For test I just changed it in the HAL, later I will see how to implement it in the driver. 

Does Zephyr have any infrastructure to enable/disable multicast support ?

pfalcon commented:
@lowlander : See https://gerrit.zephyrproject.org/r/#/c/12993/ , https://gerrit.zephyrproject.org/r/#/c/12994/ . So, it's a known issue, waiting enough in the queue by now, and first attempt at solving which failed due to differing opinions. If you start on that, feel free to keep me in loop, I'll be happy to at least offer testing on frdm_k64f.

lowlander committed:
driver: eth_stm32_hal: disable hardware multicast filtering

Until Zephyr has infrastructure to enable/disable the
reception of multicast frames we disable the hardware
multicast frame filter completly and pass all multicast
frames to the upper layer and let that deal with them.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander commented:
@pfalcon For now I just disable hardware multicast filtering in the driver. I do not yet feel confident enough  to start something where others already failed due to different opinions ;-)

pfalcon commented:
> For now I just disable hardware multicast filtering in the driver.

This is exactly the workaround applied to frdm_k64f 4-5 months ago (well, there promiscuous mode is enabled). So, unless someone (better few) will go to fix it properly, people will go thru the pain (or fun) to rediscover it again, that's all I can say. (It's definitely in my list to tackle, as commented even in those gerrit patches, still didn't get to it :-( ).

lowlander commented:
@pfalcon Zephyr is still a young project, it is to be expected things are still in flux. And just look at the fights (uhhh I mean discussions) there sometimes are on the Linux mailing list over how things should be done.

lowlander committed:
driver: eth_stm32_hal: make phy address configurable

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -0,0 +1,417 @@
+	 * at least four TX_CLK/RX_CLK clock cycles
+	 */
+	tmp = heth->Instance->MACFFR;
+	HAL_Delay(ETH_REG_WRITE_DELAY);
```
Can't we use zephyr kernel API here?

lowlander replied:
I'll lookup into it, that delay might not even be needed, I'll double check the reference manual how big it should be.

erwango commented on drivers/pinmux/stm32/pinmux_stm32f4.h:
```diff
@@ -15,10 +15,14 @@
 #define STM32F4_PINMUX_FUNC_PA0_PWM2_CH1	STM32_PINMUX_FUNC_ALT_1
 #define STM32F4_PINMUX_FUNC_PA0_UART4_TX	STM32_PINMUX_FUNC_ALT_8
 
+#define STM32F4_PINMUX_FUNC_PA1_ETH		STM32_PINMUX_FUNC_ALT_11
```
For this file, can you group and sort per pin (then sort per increasing order of ALT) so it's a bit easier to read and check for inconsistencies/duplicates?

lowlander replied:
Played catch up with that file a few times, I'll resort it.

erwango reviewed (changes requested):
Globally looks fine.
I might have missed something, but I didn't see a patch where you build HAL eth driver (in /ext/hal/st/stm32cube/Kbuild). Is there a separate PR for that?

lowlander committed:
samples: net: http_server: Add olimex_stm32_e407 project config file

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander commented:
@erwango  HAL driver is in the first commit, almost the last line.

lowlander commented:
@jukkar , @erwango any more wishes before an initial merge can be done?

erwango commented:
@lowlander: thanks, sorry for having missed it in first place.
I don't have other comments then.

jukkar reviewed (approved):
LGTM, approved
